File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/DiscussionLogic.kt
package com.lextures.android.core.lms

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray

data class DiscussionNestedPost(
    val post: DiscussionPost,
    val depth: Int,
)

object DiscussionLogic {
    fun nestPosts(posts: List<DiscussionPost>): List<DiscussionNestedPost> {
        val byParent = posts.groupBy { it.parentPostId }.mapValues { (_, list) ->
            list.sortedBy { it.createdAt }
        }
        val output = mutableListOf<DiscussionNestedPost>()
        fun walk(parentId: String?, depth: Int) {
            for (child in byParent[parentId].orEmpty()) {
                output += DiscussionNestedPost(child, depth)
                walk(child.id, depth + 1)
            }
        }
        walk(null, 0)
        return output
    }

    fun sortThreads(threads: List<DiscussionThreadSummary>): List<DiscussionThreadSummary> =
        threads.sortedWith { lhs, rhs ->
            when {
                lhs.isPinned != rhs.isPinned -> if (lhs.isPinned) -1 else 1
                else -> rhs.updatedAt.compareTo(lhs.updatedAt)
            }
        }

    fun authorLabel(authorId: String, viewerId: String?, youLabel: String): String {
        if (authorId == viewerId) return youLabel
        val trimmed = authorId.trim()
        return if (trimmed.length <= 8) trimmed else trimmed.take(8)
    }

    fun canReply(thread: DiscussionThreadDetail, viewerIsStaff: Boolean): Boolean =
        !thread.isLocked || viewerIsStaff

    fun canDeletePost(post: DiscussionPost, viewerId: String?): Boolean =
        viewerId != null && post.authorId == viewerId

    fun isBodyEmpty(text: String): Boolean = text.trim().isEmpty()

    fun plainText(body: JsonElement): String = extractText(body).trim()

    fun encodeBody(text: String): JsonElement {
        val trimmed = text.trim()
        val lines = if (trimmed.isEmpty()) listOf("") else trimmed.split('\n')
        return buildJsonObject {
            put("type", "doc")
            putJsonArray("content") {
                lines.forEach { line ->
                    add(
                        buildJsonObject {
                            put("type", "paragraph")
                            putJsonArray("content") {
                                add(
                                    buildJsonObject {
                                        put("type", "text")
                                        put("text", line)
                                    },
                                )
                            }
                        },
                    )
                }
            }
        }
    }

    private fun extractText(value: JsonElement): String =
        when (value) {
            is JsonPrimitive -> value.contentOrNull.orEmpty()
            is JsonArray -> value.joinToString("") { extractText(it) }
            is JsonObject -> {
                value["text"]?.jsonPrimitive?.contentOrNull
                    ?: value["content"]?.let { content ->
                        val separator =
                            if (value["type"]?.jsonPrimitive?.contentOrNull == "doc") "\n" else ""
                        when (content) {
                            is JsonArray -> content.joinToString(separator) { extractText(it) }
                            else -> extractText(content)
                        }
                    }
                    ?: ""
            }
            else -> ""
        }
}

File: clients/ios/Lextures/Core/LMS/DiscussionLogic.swift
import Foundation

// MARK: - Course discussions (M7.1)

struct DiscussionNestedPost: Identifiable, Equatable {
    let post: DiscussionPost
    let depth: Int
    var id: String { post.id }
}

enum DiscussionLogic {
    static func nestPosts(_ posts: [DiscussionPost]) -> [DiscussionNestedPost] {
        var byParent: [String?: [DiscussionPost]] = [:]
        for post in posts {
            byParent[post.parentPostId, default: []].append(post)
        }
        for key in byParent.keys {
            byParent[key]?.sort { $0.createdAt < $1.createdAt }
        }

        var output: [DiscussionNestedPost] = []
        func walk(parentId: String?, depth: Int) {
            for child in byParent[parentId] ?? [] {
                output.append(DiscussionNestedPost(post: child, depth: depth))
                walk(parentId: child.id, depth: depth + 1)
            }
        }
        walk(parentId: nil, depth: 0)
        return output
    }

    static func sortThreads(_ threads: [DiscussionThreadSummary]) -> [DiscussionThreadSummary] {
        threads.sorted { lhs, rhs in
            if lhs.isPinned != rhs.isPinned { return lhs.isPinned && !rhs.isPinned }
            return lhs.updatedAt > rhs.updatedAt
        }
    }

    static func authorLabel(authorId: String, viewerId: String?) -> String {
        if authorId == viewerId {
            return L.text("mobile.discussions.authorYou")
        }
        let trimmed = authorId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.count > 8 else { return trimmed }
        return String(trimmed.prefix(8))
    }

    static func canReply(thread: DiscussionThreadDetail, viewerIsStaff: Bool) -> Bool {
        !thread.isLocked || viewerIsStaff
    }

    static func canDeletePost(post: DiscussionPost, viewerId: String?) -> Bool {
        guard let viewerId else { return false }
        return post.authorId == viewerId
    }

    static func isBodyEmpty(_ text: String) -> Bool {
        text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    static func plainText(from bodyJSON: Data) -> String {
        guard !bodyJSON.isEmpty,
              let root = try? JSONSerialization.jsonObject(with: bodyJSON) else { return "" }
        return extractText(from: root).trimmingCharacters(in: .whitespacesAndNewlines)
    }

    static func encodeBody(text: String) throws -> Data {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        let lines = trimmed.isEmpty ? [""] : trimmed.components(separatedBy: "\n")
        let paragraphs: [[String: Any]] = lines.map { line in
            [
                "type": "paragraph",
                "content": [["type": "text", "text": line]],
            ]
        }
        let doc: [String: Any] = ["type": "doc", "content": paragraphs]
        return try JSONSerialization.data(withJSONObject: doc)
    }

    static func emptyBodyJSON() -> Data {
        (try? encodeBody(text: "")) ?? Data()
    }

    private static func extractText(from value: Any) -> String {
        if let string = value as? String { return string }
        if let number = value as? NSNumber { return number.stringValue }
        if let array = value as? [Any] {
            return array.map { extractText(from: $0) }.joined()
        }
        if let dict = value as? [String: Any] {
            if let text = dict["text"] as? String { return text }
            if let content = dict["content"] {
                let separator = (dict["type"] as? String) == "doc" ? "\n" : ""
                if let array = content as? [Any] {
                    return array.map { extractText(from: $0) }.joined(separator: separator)
                }
                return extractText(from: content)
            }
        }
        return ""
    }
}
